plugins/pixart-rf/fu-pxi-device.c

```diff
@@ -0,0 +1,515 @@
+fu_pxi_device_fw_object_create (FuPxiDevice *self, FuChunk *chk, GError **error)
+{
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+	guint8 res[FU_PXI_DEVICE_OTA_BUF_SZ] = { PXI_HID_DEV_OTA_INPUT_REPORT_ID };
+
+	/* request */
+	fu_byte_array_append_uint8 (req, PXI_HID_DEV_OTA_OUTPUT_REPORT_ID);
+	fu_byte_array_append_uint8 (req, FU_PXI_DEVICE_CMD_FW_OBJECT_CREATE);
+	fu_byte_array_append_uint32 (req, chk->address, G_LITTLE_ENDIAN);
+	fu_byte_array_append_uint32 (req, chk->data_sz, G_LITTLE_ENDIAN);
+	if (!fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0x0,
+					 req->data, req->len, error))
+		return FALSE;
+
+	/* reply */
+	if (!fu_udev_device_pread_full (FU_UDEV_DEVICE (self), 0x0,
+					res, sizeof(res), error))
+		return FALSE;
+	g_usleep (30 * 1000);
+
+	/* res seems unused */
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_device_write_payload (FuPxiDevice *self, FuChunk *chk, GError **error)
+{
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+	fu_byte_array_append_uint8 (req, PXI_HID_DEV_OTA_OUTPUT_REPORT_ID);
+	g_byte_array_append (req, chk->data, chk->data_sz);
+	return fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0x0,
+					   req->data, req->len, error);
+}
+
+static gboolean
+fu_pxi_device_write_chunk (FuPxiDevice *self, FuChunk *chk, GError **error)
+{
+	guint32 prn = 0;
+	guint16 checksum = fu_pxi_device_calculate_checksum (chk->data, chk->data_sz);
+	guint16 checksum_tmp = 0;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* send create fw object command */
+	if (!fu_pxi_device_fw_object_create (self, chk, error))
+		return FALSE;
+
+	/* write payload */
+	chunks = fu_chunk_array_new (chk->data, chk->data_sz, chk->address,
+				     0x0, FU_PXI_DEVICE_OTA_PAYLOAD_SZ);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk2 = g_ptr_array_index (chunks, i);
+		if (!fu_pxi_device_write_payload (self, chk2, error))
+			return FALSE;
+		prn++;
+		if (prn >= self->prn_threshold) {
+			guint8 opcode = 0;
+			if (!fu_pxi_device_wait_notify (self, 0x1, &opcode, NULL, error))
+				return FALSE;
+			if (opcode != FU_PXI_DEVICE_CMD_FW_WRITE) {
+				g_set_error (error,
+					     FWUPD_ERROR,
+					     FWUPD_ERROR_READ,
+					     "FwWrite opcode invalid %02x",
+					     opcode);
+				return FALSE;
+			}
+			prn = 0;
+		}
+	}
+
+	/* the last chunk */
+	if (!fu_pxi_device_wait_notify (self, 0x0, NULL, &checksum_tmp, error))
+		return FALSE;
+	g_debug ("checksum %x, table checksum %x", checksum_tmp, checksum);
+	if (checksum_tmp != checksum) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_READ,
+			     "checksum fail, got %x, expected %x",
+			     checksum_tmp,
+			     checksum);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_device_reset (FuPxiDevice *self, gsize fw_sz, GError **error)
+{
+	guint8 req[FU_PXI_DEVICE_OTA_BUF_SZ] = {
+		PXI_HID_DEV_OTA_OUTPUT_REPORT_ID,
+		FU_PXI_DEVICE_CMD_FW_MCU_RESET,
+		fw_sz, /* FIXME: this is size_t being sqeezed into uint8_t?! */
```

Comment:
@jyu99 is that right? It's a bit weird to squeeze a big number into just one byte...

Reply:
I will fix this bug , It should be a OTA_RESET in  previous

 enum ota_disconnect_reason
{
    OTA_CODE_JUMP = 1,          ///< OTA code jump
    OTA_UPDATE_DONE = 2,        ///< OTA update done
    OTA_RESET,                  ///< OTA reset
};

Reply:
Could I add previous `fu-pxi-device-ota.h`,becasue a lot of enum and structure  that  I need to use

Reply:
> It should be a OTA_RESET

That makes sense. Just add this enum at the top of fu-pxi-device.c -- you don't need the enum in any other C files...